jax/_src/custom_call.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2024 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from collections.abc import Iterable, Sequence
+from typing import Any
+
+import numpy as np
+
+from jax._src import core
+from jax._src import dtypes
+from jax._src.interpreters import mlir
+from jax._src.interpreters.mlir import ir
+
+
+def custom_call_lowering(
+  call_target_name: str,
```

Comment:
Nit: JAX does not strictly follow it, but Google Python style guide recommends 4 spaces for hanging indentation.